### ISSUE TITLE
"stale" locks lead to card reset on windows

### ISSUE THREAD

**Issue by philipWendland** [OpenSC/OpenSC#703]
Hi all,

I report this on behalf of a different user, i.e. I can hardly test this. Maybe he will report back here.

TL;DR: A long-term sc_lock in https://github.com/OpenSC/OpenSC/blob/master/src/tools/util.c#L159 leads to a card reset after 5 seconds when entering the user PIN in Windows 8.1 (According to the user, it is fine in W7).

https://github.com/OpenSC/OpenSC/blob/master/src/tools/util.c#L159 calls sc_lock which is maintained over a long time. The duration of the lock includes the time at which the user PIN is entered when using `pkcs15-init --create-pkcs15`.

sc_lock calls SCardBeginTransaction. See the following documentation of this function:
```
Remarks

If a transaction is held on the card for more than five seconds with no operations happening on that card, then the card is reset. 
```
https://msdn.microsoft.com/de-de/library/windows/desktop/aa379469%28v=vs.85%29.aspx
I.e. the reset after 5 seconds is documented.

Subsequently, the following will happen when OpenSC tries to send the next APDU:
```
2016-02-04 19:25:23.081 Dell Smart Card Reader Keyboard 0:SCardTransmit/Control failed: 0x80100068
```

This error code means SCARD_W_RESET_CARD.

I wonder why the call to sc_lock in https://github.com/OpenSC/OpenSC/blob/master/src/tools/util.c#L159 is present. Does anyone see a reason for this?

**Comment by martinpaljak** [comment]
This is a real annoyance. Either one needs to re-do the code to support the MS concept of only short-lived transactions or if uninterrupted access is expected during a longer period (for whatever reasons) then exclusive access can be used. Check `connect_exclusive` in opensc.conf for the last option.

**Comment by frankmorgner** [comment]
Though MS's decision on this can be debated http://permalink.gmane.org/gmane.comp.encryption.opensc.devel/16416, I think that `sc_lock()` should not be used in `util_connect_card()`. Locking was moved from the application level (e.g. opensc-explorer) into `utils.c` in beb86f351b05bdf36668887301654de86d19af35 without explanation without ever calling `sc_unlock()` again.

However, we still need `sc_lock()` somewhere to avoid accessing a card within a bad state. For example in `pkcs15-init` you don't want your card to be read out while the PKCS#15 metadata is not yet fully written.

**Comment by philipWendland** [comment]
> For example in pkcs15-init you don't want your card to be read out while the PKCS#15 metadata is not yet fully written.

This is the problem here. I think the PIN is asked for via a callback, and the writing of the PKCS#15 files is already in progress at this time.

Or did you mean locking "per object" (i.e. lock the card, write PIN and metadata, unlock and continue like that for other objects)?

**Comment by frankmorgner** [comment]
What state is considered "consistent" should be decided within the application. E.g. with `pkcs15-init` you may generate a private key on the card without locking, but I guess you want to lock the card when writing all PKCS#15 objects that are referencing themselves (PRKDF, AODF, ...). And for sure you don't need to lock the card on connecting when using `opensc-explorer`.

**Comment by frankmorgner** [comment]
I opened https://github.com/OpenSC/OpenSC/issues/750 to track this issue

**Comment by szszszsz** [comment]
Hi!
I wanted to revive the discussion about this issue. Would my PR be sufficient as a quick fix?
We have tested it on Windows 8 with command:
```
sc-hsm-tool.exe --create-dkek-share dkek-share-1.pbe
```
and it looks like it worked (whereas without it the card was reset just after entering password with 7 seconds pause on same test environment). 
Please check would any additional changes be needed to maintain stability. If possible please test it. If there is any other way to avoid card being reset on Windows 8+ after delay please let me know.
Artifacts rebased on stable release `0.16` are available here:
https://ci.appveyor.com/project/szszszsz/opensc/build/0.16.0.3
